spec/controllers/users_controller_spec.rb

```diff
@@ -0,0 +1,85 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe UsersController, type: :controller do
+  let!(:user) { create(:user) }
+  let!(:valid_params) { attributes_for :user }
+  let!(:invalid_params) { { first_name: ' ', last_name: ' ' } }
+
+  describe 'GET#index' do
+    it 'assigns users and renders template' do
+      get :index
+      expect(response).to have_http_status(:success)
+      expect(response).to render_template('index')
+    end
+  end
+
+  describe 'GET#show' do
+    before do
+      get :show, params: { id: user.id }
+    end
+
+    it 'returns success and assigns user' do
+      expect(response).to have_http_status(:success)
+      expect(assigns(:user)).to eq(user)
+    end
+  end
+
+  describe 'GET#edit' do
+    before do
+      get :edit, params: { id: user.id }
+    end
+
+    it 'returns http success and assign user' do
+      expect(response).to have_http_status(:success)
+      expect(assigns(:user)).to eq(user)
+    end
+  end
+
+  describe 'PUT#update' do
+    context 'with valid params' do
+      before do
+        put :update, params: { id: user.id,
+                               user: valid_params.merge!(first_name: 'Example',
+                                                         last_name: 'Test',
+                                                         email: 'example@email.com') }
+      end
+
+      it 'assigns the user' do
+        expect(assigns(:user)).to eq(user)
+        expect(response).to have_http_status(:redirect)
+        expect(response).to redirect_to(user_path(user))
+      end
+
+      it 'updates user attributes' do
+        user.reload
+        expect(user.first_name).to eq(valid_params[:first_name])
+        expect(user.last_name).to eq(valid_params[:last_name])
+        expect(user.email).to eq(valid_params[:email])
```

Comment:
1. `valid_params` are `attributes_for(:user)`, but `params` are `{ first_name: 'Example', ... }`. in case of a successful update, it's expected that the new user first name should be exactly the same, as passed to `params`. your `params` are different to `valid_params`.
it should fail

2. 
```
expect(user.attributes.symbolize_keys).to include(valid_params.slice(:first_name, :last_name, :email))
```

Reply:
done

Reply:
done